### PR TITLE
:sparkles: add kubebuilder/tooling and controller setup to use binaries from the project bin (go/v3-alpha)

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description**
- add hack/tooling to setup the binaries in the bin project when the KUBEBUILDER_ASSETS is not setup
- update the quick start to make clear that the current steps to get the binaries still required until the CLI version   `v2.3.1`

**Motivation**
-  provide a solution that allows users run and debug/troubleshooting the tests, that not requires sudo permissions to setup an env to develop the projects, and provide a code easier to maintain and to read. 
- Closes : https://github.com/kubernetes-sigs/kubebuilder/issues/1732
- Closes: https://github.com/kubernetes-sigs/kubebuilder/issues/1892

### CODE REVIEW

File: docs/book/src/quick-start.md
Comment (estroz) [inline docs/book/src/quick-start.md:30]: Don't you need to do the steps below regardless of kubebuilder version, since the kubebuiler binary itself is in that tarball?

Comment (camilamacedo86) [reply docs/book/src/quick-start.md:30]: These steps are required currently/today. So, we need to keep the steps to let the users know what is required to do for versions until the current one.

**Review by gabbifish — APPROVED** [review]
/approve

I ran `make test` as well and all your introduced changes look good! :)

File: pkg/plugin/v2/scaffolds/init.go
Comment (DirectXMan12) [inline pkg/plugin/v2/scaffolds/init.go:45]: We almost certainly shouldn't be pinning at `master` -- this is gonna break

Comment (camilamacedo86) [inline pkg/plugin/v2/scaffolds/init.go:45]: we just spoke in the meeting. Shows that we have a release already for that. So, we need ping the tag. 
also, we cannot merge point for the master.

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
comment inline, plus as was pointed out in the meeting, we should verify the downloaded tarball.

**Comment by DirectXMan12** [comment]
/hold

re: pointing at "master", we can bcckport the script to a stable version if need-be

File: pkg/plugin/v2/scaffolds/internal/templates/makefile.go
Comment (DirectXMan12) [inline pkg/plugin/v2/scaffolds/internal/templates/makefile.go:78]: this should be a separate target, and should check for existing files.

File: testdata/project-v2/Makefile
Comment (camilamacedo86) [inline testdata/project-v2/Makefile:21]: For v2, we need to improve that to keep the same bahaviour. 
If the bins already exist there then we should use instead of download.

File: pkg/plugin/v2/scaffolds/init.go
Comment (camilamacedo86) [reply pkg/plugin/v2/scaffolds/init.go:45]: I just checked that we have a new version to bump. https://github.com/kubernetes-sigs/controller-runtime/blob/v0.6.3/hack/setup-envtest.sh

Comment (DirectXMan12) [reply pkg/plugin/v2/scaffolds/init.go:45]: we may also just want to have a separate script of kubebuilder or something that we scaffold out, or a go module that we can `go get` that does the downloading.

Comment (DirectXMan12) [reply pkg/plugin/v2/scaffolds/init.go:45]: e.g. have a separate module in `kubebuilder/tooling` (so it has its own deps), and then we can `go get`.

Comment (camilamacedo86) [reply pkg/plugin/v2/scaffolds/init.go]: I really liked the idea of the tooling and I also agree that shows better we do not need to create a dependency to controller runtime to just download the bins.

File: docs/book/src/quick-start.md
Comment (estroz) [reply docs/book/src/quick-start.md:30]: Agreed. I am saying that you'll need to run these steps no matter what version you're using. The tool binaries haven't yet been removed from the release tarball, and the release tarball contains the kubebuiler binary itself.
```suggestion
To add `kubebuilder` and [envtest][envtest] binaries to your path: 
```

File: pkg/plugin/v3/scaffolds/internal/templates/makefile.go
Comment (DirectXMan12) [inline pkg/plugin/v3/scaffolds/internal/templates/makefile.go:84]: why not just scaffold out `/.tooling` with a go.mod,
then this becomes `cd $PROJECT/.tooling; go run sigs.k8s.io/kubebuilder/tooling` (FWIW, I'm not sure we need do the whole "is there a predefined helper binary" dance here -- we can just build the right nobs into the helper binary)

File: tooling/main.go
Comment (DirectXMan12) [inline tooling/main.go:26]: please don't introduce yet another logging library, especially one that's deprecated.  Use either logr with zapr, or klog v2

File: tooling/main.go
Comment (DirectXMan12) [inline tooling/main.go:36]: maybe want to consider moving this to a separate branch so that it's easy to bump the tooling (can maybe just put this code on the existing tooling branch)

File: tooling/main.go
Comment (DirectXMan12) [inline tooling/main.go:42]: check if `KUBEBUILDER_ASSETS` is set too

File: tooling/main.go
Comment (DirectXMan12) [inline tooling/main.go:52]: have a `bin path` option to configure this.

also, we might (not quite sure, what do other folks think?) want to consider installing into $GOBIN with a version suffix, then symlinking to bin, or something, like

```
$GOBIN/kubectl-1.15.6
$PROJECT/.tooling/bin/kubectl -> $GOBIN/kubectl-1.15.6
```

That way, if you're working on a lot of different kb projects (which some folks do), you don't end up with 15 copies of kube-apiserver floating around.

File: tooling/main.go
Comment (DirectXMan12) [inline tooling/main.go:64]: :-/  Any reason not to just natively use the Go http client for this?  It seems a little weird to shell out to curl

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go
Comment (camilamacedo86) [reply pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go:179]: IHMO yes. But this change required to get done in the controller-runtime see: https://github.com/kubernetes-sigs/controller-runtime/pull/1214

File: tooling/main.go
Comment (camilamacedo86) [reply tooling/main.go:64]: no reason at all. Good point.

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go
Comment (camilamacedo86) [reply pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go]: Also, I centralized the code in a hack/utils so far. (new scaffold)

File: pkg/plugin/v3/scaffolds/internal/templates/makefile.go
Comment (camilamacedo86) [reply pkg/plugin/v3/scaffolds/internal/templates/makefile.go]: Done.

File: tooling/main.go
Comment (camilamacedo86) [reply tooling/main.go]: Done.

File: tooling/main.go
Comment (camilamacedo86) [reply tooling/main.go]: IMO the default assets path in controller-runtime ( https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/envtest/server.go#L57 ) should be `$GOBIN/` as you suggested instead of `"/usr/local/kubebuilder/bin"` which requires sudo permissions. And then, the project tooling set up in the project when it is not set up with the default path. In this way, users are also able to have diff versions per project as they wish.

Comment (camilamacedo86) [reply tooling/main.go]: Raised the issue https://github.com/kubernetes-sigs/controller-runtime/issues/1234 in the controller-runtime

**Comment by estroz** [comment]
Is there a significant advantage to scaffolding code over downloading the controller-runtime script to download envtest binaries? The method proposed by this PR feels clunky. I'd prefer either an easier way to download envtest binaries from controller-runtime using a script, or exposing a Go library in controller-runtime to do so. In general I'd like to see dependencies manage their own installation/usage as much as possible so kubebuilder doesn't have to implement workarounds.

**Comment by camilamacedo86** [comment]
Hi @estroz, 

> Is there a significant advantage to scaffolding code over downloading the controller-runtime script to download envtest binaries? The method proposed by this PR feels clunky. I'd prefer either an easier way to download envtest binaries from controller-runtime using a script, or exposing a Go library in controller-runtime to do so. In general I'd like to see dependencies manage their own installation/usage as much as possible so kubebuilder doesn't have to implement workarounds.

- We added a script to controller-runtime to do it, however, I am afraid that it is not really of the controller-runtime responsibility either. So, ` In general, I'd like to see dependencies manage their own installation/usage` then, it would be in Kubeuilder and not controller-runtime. 
- A few advantages of a scaffold approach here in pov are:  to allow users to customize it as they wish, give more visibility over how it works and keep the implementation where it is required/used.
 
**Notes** Also, see that it was a suggestion made by @DirectXMan12  in its comments as well. See the changes made in the suite_test.go scaffold will no longer be required after we merge  kubernetes-sigs/controller-runtime#1214.

**Comment by estroz** [comment]
>I am afraid that it is not really of the controller-runtime responsibility either.

envtest is a controller-runtime library that requires k8s binaries to run. Perhaps kubebuilder can ease setup a bit, but controller-runtime should provide an easy mechanism to at least install these binaries locally. #686 is already suggesting they be released separately.

>A few advantages of a scaffold approach here in pov are: to allow users to customize it as they wish, give more visibility over how it works and keep the implementation where it is required/used.

The same benefits would be achieved by a controller-runtime library, with less maintenance cost in kubebuilder and the benefit of module versioning.

**Comment by camilamacedo86** [comment]
Hi @estroz,

> #686 is already suggesting they be released separately from kubebuilder, so IMO we should take that idea a step further and have controller-runtime release them.

I just would like to make an observation that this pr helps achieve the goal of  #686 and not the opposite. See that any way/approach and language that we decide to use to download bin from the cloud solve that no matter where. 

Regards how we will solve this issue, see that what you are suggesting was exactly how it was done before and what was reject in this pr at the first moment. So, I only think that we need to reach a consensus before continue.

**Comment by DirectXMan12** [comment]
I'd prefer a runnable Go program to a shell script, I think (there's just so many things that can go wrong in a shell script, with a `go run` invocation orgs can use a custom module proxy to ensure that they're getting what they expect, etc).

That being said, I think I put on one of these scripts that we should probably install to GOBIN, and then symlink the particular version to bin/whatever or something, so we still share versions when appropriate -- having 27 copies of the same binary in `~` is not ok.

If we want that to live in a separate module in controller-runtime, that's fine.

If we want to migrate the kubernetes release branch to controller-runtime, and re-target the cloud build scripts, that's also probably fine (I don't really care strongly either way).

**Comment by estroz** [comment]
>I'd prefer a runnable Go program to a shell script

In general I agree.

>That being said, I think I put on one of these scripts that we should probably install to GOBIN, and then symlink the particular version to bin/whatever or something, so we still share versions when appropriate -- having 27 copies of the same binary in ~ is not ok.

The library API could have "desired version" parameters that gets compared to the versions of GOBIN/PATH binaries, and will install any of those binaries locally if its version does not match the desired version.

>If we want that to live in a separate module in controller-runtime, that's fine.

I'd prefer that, so there's a canonical way of installing envtest deps.

>If we want to migrate the kubernetes release branch to controller-runtime, and re-target the cloud build scripts, that's also probably fine (I don't really care strongly either way).

Ultimately I'd like to do that as well, especially if c-r has the installation library you mentioned above.

**Comment by camilamacedo86** [comment]
Hi @estroz @DirectXMan12, 

This change shows make sense only for v3-alpha so I updated it. 

Regards the suggestions made by @estroz for the binaries be generated from controller-runtime instead of kubebuilder it is out of the scope what we are trying to achieve here. 

Regards the tooling to install the binaries be provided by controller-runtime shows harder we keep it maintained. However, if we reach this consensus then, it ought to be made with or after the binaries be generated from there. 

In this way, I think that we should do:

- Move forward with this solution to solve the current problems in KB ( we need to provide a solution that allows users run and  debug/troubleshooting the tests, that not requires sudo permissions to setup an env to develop the projects, and provide a code easier to maintain and to read)
- Raise an issue against controller-runtime to starts to generate the binaries from there
- Copy adapted if it is required the tooling done here to controller-runtime 
- After the next controller-runtime release ith the suggestions made then, we can start to use it and change the KB code implementation.

**Comment by estroz** [comment]
Common install path and library suggestions: https://github.com/kubernetes-sigs/controller-runtime/issues/1234

**Comment by camilamacedo86** [comment]
Closing it a favour to https://github.com/kubernetes-sigs/controller-runtime/issues/1234 and https://github.com/kubernetes-sigs/controller-runtime/issues/1300 as discussed in the bug triage meeting.

**Pull request closed without merging**
